include/graphblas/algorithms/hpcg/average_coarsener.hpp

```diff
@@ -0,0 +1,412 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @file average_coarsener.hpp
+ * @author Alberto Scolari (alberto.scolari@huawei.com)
+ * Utilities to build the coarsening matrix for an HPCG simulation.
+ */
+
+#ifndef _H_GRB_ALGORITHMS_AVERAGE_COARSENER
+#define _H_GRB_ALGORITHMS_AVERAGE_COARSENER
+
+#include <array>
+#include <cmath>
+#include <cstddef>
+#include <iterator>
+#include <numeric>
+#include <stdexcept>
+
+#include <graphblas/utils/multigrid/array_vector_storage.hpp>
+#include <graphblas/utils/multigrid/linearized_ndim_system.hpp>
+
+
+namespace grb {
+
+	namespace algorithms {
+
+		namespace hpcg {
+
+			// forward declaration
+			template<
+				size_t DIMS,
+				typename CoordType,
+				typename ValueType
+			>
+			class AverageCoarsenerBuilder;
+
+			/**
+			 * Iterator class to generate the coarsening matrix that averages over the
+			 * elements of the finer domain corresponding to the element of the coarser
+			 * domain.
+			 *
+			 * The coarsening matrix averages \b all elements that are coarsened into
+			 * one.
+			 *
+			 * This coarsening method requires some computation but should be relatively
+			 * robust to noise or to partitioning strategies that parallelize the
+			 * smoother (usually run before coarsening).
+			 *
+			 * This iterator is random-access.
+			 *
+			 * @tparam DIMS number of dimensions
+			 * @tparam CoordType type storing the coordinates and the sizes
+			 * @tparam ValueType type of the nonzero: it must be able to represent 1 /
+			 *                   <number of finer elements per coarser elements>
+			 */
+			template<
+				size_t DIMS,
+				typename CoordType,
+				typename ValueType
+			>
+			struct AverageGeneratorIterator {
+
+				friend AverageCoarsenerBuilder< DIMS, CoordType, ValueType >;
+
+				/** Numeric type of rows */
+				typedef CoordType RowIndexType;
+
+				/** Numeric type of columns */
+				typedef CoordType ColumnIndexType;
+
+				typedef typename grb::utils::multigrid::LinearizedNDimSystem<
+						CoordType,
+						grb::utils::multigrid::ArrayVectorStorage< DIMS, CoordType >
+					> LinearSystemType;
+
+				typedef typename LinearSystemType::Iterator LinearSystemIterType;
+
+				typedef AverageGeneratorIterator< DIMS, CoordType, ValueType > SelfType;
+
+				typedef std::array< CoordType, DIMS > ArrayType;
+
+				class ValueGenerator {
+
+					friend SelfType;
+
+
+					private:
+
+						RowIndexType _i;
+
+						ColumnIndexType _j;
+
+						ValueType _value;
```

Comment:
These private members were used by the below code, while the private field declaration was put at the bottom of the class. The convention is to declare it first and then use it, so I inverted the two blocks